rfcs/framework-0040-relations.md

```diff
@@ -0,0 +1,321 @@
+- We would like a syntax that can be used both in the internal JSON document stored in the database, as well as in the external JSON used in the REST API.
+- By using an object we can easily dereference it. I.e. ``{"id": "..."}`` is dereferenced to ``{"id": "...", "key1": "val1", ...}``. The key ``id`` always exists (unlike e.g. when using JSONRef).
+- Relations should be to a persistent identifier for the related record to make the relation as stable as possible. Also, we wanted to avoid having mulit-key objects as possible persistent identifiers to ensure APIs could be as standardized as much as possible.
+- We did not want to include e.g. type information for the related record, to keep the JSON small as well as keep the type information on the field definition itself, instead of inside the referenced object.
+
+#### Persistent identifiers
+
+We imagine that the persistent identifiers depends on the related record. If the related record uses Invenio-PIDStore, it would be the persistent identifer stored in PIDStore. If the related record just uses the record's table UUID as persistent identifier, this value could be used.
+
+Persistent identifiers should not be renamed to avoid having to go update the relations in database records (i.e the persistent identifier is actually persistent!).
+
+If renaming is needed, it could be done by:
+
+- Adding a new persistent identifier.
+- Redirect the old persistent identifier to the new.
+- Change all records to use the new persistent identifier
+- Delete the old persistent identifier.
+
+#### Integrity
+
+The checking of integrity (i.e. that a related record exists), can be done via a simple primary key look up for the persistent identifier.
+
+#### Dereferencing
+
+The dereferencing mechanism should allow for fine-grained control of which fields are dereferenced. This is because the reference object may be big, and only a subset of fields is relevant to dereference.
+
+For instance, take a record which should be related:
+
+```json
+{
+    "id": "1234",
+    "a": "",
+    "b": ""
+}
+```
+
+The relation in another record would look like below:
+
+```json
+{"id": "1234"}
+```
+
+When dereferenced, it should be possible to only include the key ``a`` and not ``b``:
+
+```json
+{"id": "1234", "a": ""}
+```
+
+#### Denormalization
+
+Dereferencing a relation can be compared to denormalizing the data or caching the data. This means, that once the related record changes, the places where it was dereferenced/cached must be updated.
+
+The issues with that is:
+
+- **Cardinality**: If a related record is referenced by 1 million records, you have to go and dereference 1 million records again.
+
+- **Concurrency**: With high cardinality, dereferencing will take time, and thus during this time the system can operate with stale data. For instance a new entry may be added, that must be picked up for dereferencing to avoid having it around forever with stale data.
```

Comment:
What does "that" in "that must be picked up" refer to? 

Isn't it the case that a new entry is the *least* problematic situation? A problematic situation would be:

- Relation to bar is derefenced in a million records
- bar is changed, an asynchronous task A to dereference bar in the million records is launched
- bar is changed again, another asynchronous task B is launched
- task B finishes first on a record and then task A finishes

The asynchronous task should not be carrying the changes themselves to prevent issues. It should just tell the record to dereference again. It will pick up latest changes then.

Not clear to me what are the actual concurrency issues here... don't know that stale data is a problem as long as it is eventually updated (the previous data was fine otherwise we would have had an issue then and if it's now not fine, then there is probably a dependency issue at the root of it really...).

Reply:
The situation is the following:

1. Transaction 1 starts - record Foo v1 is changed to v2
2. Transaction 2 starts - create new record Bar with reference to Foo (v1 is read because v2 is not yet committed)
3. Transaction 1 commits - Foo is now in v2. 
4. Transaction 1 starts a scan of the ES index to find records (same index as Bar) with reference to Foo v1
5. Transaction 2 commits
6. Transaction 2 - record Bar will now reference the relation to Foo. Foo v1 will be picked up, because it's already in the cache.

Result now is:
1) Bar has Foo v1 dereferenced inside the ES index
2) Bar is not included list of records that must be dereferenced again, because the ES index scan is stable and was started before Bar was indexed.

Thus the problem is that the record Bar is never updated and you have forever stale data.

The situation that you describe above is case where the version counter fixes the issue. The referencing task, will always read the latest version, so the only problem there is that we index a million records twice. That's a bit waste of resource, but eventually lead to a consistent search index. The problem above, is that we never reach a consistent search index.

Reply:
Thanks for the explanations. I think I see the problem. So what is the solution? I see 2 possibilities, (but I am not fluent in the fundamental concurrency tools we have at our disposal) :

1- Out-of-band dereferencing updates: celery beat to dereference all relations of indexed record. (eventually consistent. Could be complementing other approaches too.)
2- Enlarging the optimistic concurrency control scope: Transaction 2 can't commit unless its related records' versions match what is in the database (i.e. if Transaction 1 committed before, then Transaction 2's version for Foo is out of date at time of commit and would be rolled back).

My understanding is that our concurrency tools are limited at ES level, but better at DB level. So perhaps other options at DB...?